sw/device/lib/testing/mock_mmio.h

```diff
@@ -0,0 +1,78 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_MOCK_MMIO_H_
+#define OPENTITAN_SW_DEVICE_LIB_TESTING_MOCK_MMIO_H_
+
+#include <stdint.h>
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+namespace mock_mmio {
+/**
+ * A MockDevice represents a mock implementation of an MMIO device.
+ *
+ * MockDevice provides two mockable member functions, representing a read and a
+ * write at a particular offset from the base address. This class can be
+ * converted into a |reg32_t| value, which, when used in |mmio.h| functions like
+ * |reg32_read()|, will map to the appropriate mock member function calls.
+ *
+ * To maintain sequencing, |ReadInternal()| and |WriteInternal()| should not be
+ * |EXPECT_CALL|'ed directly; instead, |EXPECT_READ| and |EXPECT_WRITE| should
+ * be used, instead.
+ *
+ * To use this class, |-DMOCK_MMIO| must be enabled in all translation units
+ * using |mmio.h|.
+ */
+class MockDevice {
+ public:
+  MockDevice() = default;
+
+  MockDevice(const MockDevice &) = delete;
+  MockDevice &operator=(const MockDevice &) = delete;
+  MockDevice(MockDevice &&) = delete;
+  MockDevice &operator=(MockDevice &&) = delete;
+
+  /**
+   * Converts this MockDevice into a reg32_t opaque object, which is compatible
+   * with |mmio.h| functions.
+   */
+  reg32_t Register();
+
+  MOCK_METHOD(uint32_t, ReadInternal, (ptrdiff_t offset));
+  MOCK_METHOD(void, WriteInternal, (ptrdiff_t offset, uintptr_t value));
+
+  // NOTE: This member is only public for use by macros, and should not be
+  // considered part of the public API.
+  testing::Sequence seq_;
+};
+}  // namespace mock_mmio
+
+/**
+ * Expect a read to the device |dev| at the given offset, returning the given
+ * 32-bit value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_READ(dev, offset, value)  \
```

Comment:
#1300 introduces a test fixture that provides these as methods, not macros. Please let me know what you think.

Reply:
Well, your version is forced to put `auto` in return position, which I don't think is advisable. I think the cost of being able to abuse implicit `this` to avoid naming the device in each EXPECT call is higher than the benefit it provides.

Reply:
`auto` is used only to avoid referring to a type that is internal to GoogleTest.

Reply:
Personally, I feel like that's reason enough to make it a macro.

Also, I believe that `EXPECT_CALL` uses `__FILE__` and `__LINE__`, so your version might not give the best error messages...

Reply:
Hmm, this is a very good point. It looks like it does and error messages point to `mmio_mock.` in my case, so not very helpful... Thanks for pointing this out!.